From c4a558aab2d6b213e1b862248e0b2fc19dd81cc5 Mon Sep 17 00:00:00 2001
From: Darren Birkett <darren.birkett@gmail.com>
Date: Tue, 14 Jun 2016 11:43:12 +0000
Subject: [PATCH] Ensure all hosts have a pip.conf

Previously we were relying on some other side effects to get a
pip.conf file to be put onto the hosts:

- compute hosts got it because of the nova role
- infra hosts sometimes got it because of beaver/other roles

This commit ensures that we explicitly install a pip.conf for
the version that we are deploying by adding a playbook to run
the pip-lockdown role. Additionally, we remove any lingering
link files from previous versions to ensure that pip.conf
contains only links for the new version being installed

(cherry picked from commit 3afa31a130690b95ca1179dda1621638328ec575)
---
 rpcd/playbooks/pip-lockdown.yml | 35 +++++++++++++++++++++++++++++++++++
 scripts/deploy.sh               |  5 +++++
 scripts/upgrade.sh              |  8 ++++++++
 3 files changed, 48 insertions(+)
 create mode 100644 rpcd/playbooks/pip-lockdown.yml

diff --git a/rpcd/playbooks/pip-lockdown.yml b/rpcd/playbooks/pip-lockdown.yml
new file mode 100644
index 0000000..8f9e426
--- /dev/null
+++ b/rpcd/playbooks/pip-lockdown.yml
@@ -0,0 +1,35 @@
+---
+# Copyright 2016, Rackspace US, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# The purpose here is to allow for the environment to update/build the
+# python wheel files from the CURRENT release, as set in the rpc_release
+# variable.
+- name: Clean up pip release files from previous versions
+  hosts: hosts:all_containers
+  user: root
+  gather_facts: False
+  tasks:
+    - name: remove pip release directory
+      file:
+        path: /root/.pip/links.d/
+        state: absent
+      tags:
+        - pip-link-cleanup
+
+- name: Place pip.conf on hosts
+  hosts: hosts
+  user: root
+  roles:
+    - { role: "pip_lock_down" }
diff --git a/scripts/deploy.sh b/scripts/deploy.sh
index 097d6a9..5ebc6e5 100755
--- a/scripts/deploy.sh
+++ b/scripts/deploy.sh
@@ -130,6 +130,11 @@ if [[ "${DEPLOY_OA}" == "yes" ]]; then
   # setup the hosts and build the basic containers
   run_ansible setup-hosts.yml
 
+  # ensure correct pip.conf
+  pushd ${RPCD_DIR}/playbooks/
+    run_ansible pip-lockdown.yml
+  popd
+
   if [[ "$DEPLOY_CEPH" == "yes" ]]; then
     pushd ${RPCD_DIR}/playbooks/
       run_ansible ceph-all.yml
diff --git a/scripts/upgrade.sh b/scripts/upgrade.sh
index 57084ef..a6b5050 100755
--- a/scripts/upgrade.sh
+++ b/scripts/upgrade.sh
@@ -40,6 +40,14 @@ ansible-galaxy install --role-file=/opt/rpc-openstack/ansible-role-requirements.
 cd ${RPCD_DIR}/playbooks
 openstack-ansible -i "localhost," patcher.yml
 
+# Make sure ~/.pip does not exist on repo containers, it prevents the
+# repo_build role from installing new pip packages because the version
+# are not on the repo server
+ansible repo_container -a 'rm -rf /root/.pip'
+
+# ensure correct pip.conf
+openstack-ansible pip-lockdown.yml
+
 # Do the upgrade for openstack-ansible components
 cd ${OA_DIR}
 echo 'YES' | ${OA_DIR}/scripts/run-upgrade.sh
-- 
2.8.3

